### PR TITLE
refactor to support multi-part upload (GDEV-978)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
A major refactoring with many changes. Sorry.

Please don't squash and merge but do a regular merge commit.
The history has been cleaned.

### CODE REVIEW

**Comment by KerstenBreuer** [comment]
Sorry, the chassis lib dependency is not yet available. I will restart the tests in a couple of minutes.

File: tests/fixtures/s3.py [hunk 1]
```diff
@@ -15,28 +15,55 @@
...
+        """Populate the storage with file objects."""
+
+        populate_storage(self.storage, bucket_fixtures=[], object_fixtures=file_objects)
+
+
+@pytest.fixture
+def s3_fixture() -> Generator[S3Fixture, None, None]:
+    """Pytest fixture for tests depending on the ObjectStorageS3 DAO."""
+
+    with LocalStackContainer(image="localstack/localstack:0.14.2").with_services(
```
Comment (mephenor) [inline tests/fixtures/s3.py:55]: Should this be hardcoded here, i.e. can we get the current version from another place where it's already defined?

File: tests/test_api_journey.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import json
+from datetime import datetime
+from typing import Literal
+
+from fastapi import status
+from ghga_message_schemas import schemas
+from ghga_service_chassis_lib.utils import exec_with_timeout
+
+from tests.fixtures.example_data import EXAMPLE_FILES
+from tests.fixtures.joint import *  # noqa: 403
```
Comment (mephenor) [inline tests/test_api_journey.py:28]: Shouldn't wildcard imports be avoided in general?

File: tests/test_file_service.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    # construct service and dependencies:
+    upload_attempt_dao = PsqlUploadAttemptDAO(db_url=psql_fixture.config.db_url)
+    file_metadata_dao = PsqlFileMetadataDAO(db_url=psql_fixture.config.db_url)
+    file_metadata_service = FileMetadataServive(
+        file_metadata_dao=file_metadata_dao,
+        upload_attempt_dao=upload_attempt_dao,
+    )
+
+    # create file metadata:
+    file_metadata_service.upsert_multiple(EXAMPLE_FILES)
```
Comment (mephenor) [inline tests/test_file_service.py:83]: What does `upsert_multiple` mean, some concatenation of upload and insert?

File: tests/test_psql_file.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        with pytest.raises(FileMetadataNotFoundError):
+            _ = file_metadata_dao.get(expected_file.file_id)
+
+
+def test_create_file_metadata(psql_fixture: PsqlFixture):  # noqa: F811
+    """Test registering not existing file metadata."""
+    expected_file = EXAMPLE_FILE
+
+    with PsqlFileMetadataDAO(db_url=psql_fixture.config.db_url) as file_metadata_dao:
+        file_metadata_dao.upsert(expected_file)
```
Comment (mephenor) [inline tests/test_psql_file.py:52]: Same as above with `upsert`

File: tests/test_rest.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+"""Test edge cases of the fastapi_ adapter not covered by `test.test_api_journey`."""
+
+import json
+
+import pytest
+from fastapi import status
+
+from tests.fixtures.example_data import EXAMPLE_FILE, EXAMPLE_UPLOADS
+from tests.fixtures.joint import *  # noqa: 403
```
Comment (mephenor) [inline tests/test_rest.py:24]: As above

File: tests/test_upload_service.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+import pytest
+from ghga_message_schemas import schemas
+
+from tests.fixtures.example_data import (
+    EXAMPLE_FILE,
+    EXAMPLE_STORAGE_OBJECT,
+    EXAMPLE_UPLOAD,
+)
+from tests.fixtures.joint import *  # noqa: 403
```
Comment (mephenor) [inline tests/test_upload_service.py:26]: As above

File: ucs/domain/models.py [hunk 7]
```diff
@@ -17,58 +17,87 @@
...
 
-class FileInfoPatchState(BaseModel):
+class UploadAttempt(BaseModel):
     """
-    A model containing all the metadata needed to perform a patch on an orm field
+    A model containing details on an upload attempt for a specific File.
     """
 
-    state: UploadState = UploadState.REGISTERED
+    upload_id: str
```
Comment (mephenor) [inline ucs/domain/models.py:52]: Just for consistency's sake, shouldn't this also be a Field?

File: ucs/domain/models.py [hunk 8]
```diff
@@ -17,58 +17,87 @@
...
+
+        orm_mode = True
+        title = "Multi-Part Upload Details"
 
-class FileInfoExternal(FileInfoPatchState):
+
+class FileMetadata(BaseModel):
     """
-    A model containing all the metadata needed to pass it on to other microservices
+    A model containing basic metadata on a file.
```
Comment (mephenor) [inline ucs/domain/models.py]:
```suggestion
    A model containing basic file metadata.
```

File: ucs/domain/models.py [hunk 9]
```diff
@@ -17,58 +17,87 @@
...
         orm_mode = True
+        title = "Basic File Metadata"
 
 
-class FileInfoInternal(FileInfoExternal):
+class FileMetadataWithUpload(FileMetadata):
     """
-    A model containing all the metadata submitted for one file from the metadata service
-    with the new_study_created topic.
+    A model containing basic metadata on a file plus information on the current
```
Comment (mephenor) [inline ucs/domain/models.py]:
```suggestion
    A model containing basic file metadata and information on the current
```

File: ucs/domain/part_calc.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    while True:
+        yield part_size
+
+        # double the part size after every iteration:
+        part_size = part_size * 2
+
+
+# pylint: disable=inconsistent-return-statements
+def calculate_part_size(file_size: int) -> int:  # type: ignore
+    # MyPy and pyling complains that the function does not return, however, they do not
```
Comment (mephenor) [inline ucs/domain/part_calc.py]:
```suggestion
    # MyPy and pylint complain that the function does not return, however, they do not
```

File: ucs/domain/upload_service.py [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+        with self._object_storage as storage:
+            if not storage.does_bucket_exist(bucket_id=self._s3_inbox_bucket_id):
+                storage.create_bucket(self._s3_inbox_bucket_id)
+
+    def _get_upload_if_status(
+        self, upload_id: str, status: models.UploadStatus
+    ) -> models.UploadAttempt:
+        """Makes sure that the upload with the given ID exists and that its current
+        status matches the specified status. If that is the case, the upload is return. Otherwise
+        an UploadStatusMissmatchError is raise."""
```
Comment (mephenor) [inline ucs/domain/upload_service.py]:
```suggestion
        an UploadStatusMissmatchError is raised."""
```

File: ucs/domain/upload_service.py [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+
+    def _cancel_with_final_status(
+        self, *, upload_id: str, final_status: models.UploadStatus
+    ) -> None:
+        """
+        Cancel the multi-part upload with the given ID and set the specified final
+        status.
+        """
+
+        upload = self._get_upload_if_status(
```
Comment (mephenor) [inline ucs/domain/upload_service.py:115]: What's the if in `_get_upload_if_status` ?

File: ucs/domain/upload_service.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+            except ObjectNotFoundError as error:
+                # This is unexpected. Thus setting the status of the upload attempt to
+                # failed and raise.
+                self._cancel_with_final_status(
+                    upload_id=latest_upload.upload_id,
+                    final_status=models.UploadStatus.FAILED,
+                )
+                raise StorageAndDatabaseOutOfSyncError(
+                    problem=(
+                        f"Trying clear the upload with ID {latest_upload.upload_id} (final status:"
```
Comment (mephenor) [inline ucs/domain/upload_service.py]:
```suggestion
                        f"Trying to clear the upload with ID {latest_upload.upload_id} (final status:"
```

File: ucs/domain/upload_service.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+            # the storage)
+            with self._upload_attempt_dao as ua_dao:
+                try:
+                    ua_dao.create(upload)
+                except:
+                    # One source of error might be that an upload with the given ID
+                    # already exists. In that case the assumption that the object
+                    # storage assigns unique IDs is violated. However, at this stage
+                    # there is nothing we can do to handel this exception.
+                    storage.abort_multipart_upload(
```
Comment (mephenor) [inline ucs/domain/upload_service.py:247]: Can this fail here?

File: ucs/domain/interfaces/inbound/upload_service.py [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+        message = (
+            f"The upload with ID {upload_id} must be in '{expected_status}' state to"
+            + " perform the requested action, however, its current state is:"
+            + str(current_status)
+        )
+        super().__init__(message)
+
+
+class ExistingActiveUploadError(RuntimeError):
+    """Thrown when trying to create a new upload while there is another upload active."""
```
Comment (mephenor) [inline ucs/domain/interfaces/inbound/upload_service.py]:
```suggestion
    """Thrown when trying to create a new upload while another upload is already active."""
```

File: ucs/domain/interfaces/inbound/upload_service.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+class UploadCompletionError(RuntimeError):
+    """Thrown when the confirmation of an upload attempt failed."""
+
+    def __init__(self, *, upload_id: str, reason: str):
+        self.reason = reason
+        self.upload_id = upload_id
+        message = (
+            f"The confirmation of the upload attempt with ID {upload_id} failed."
+            + " The upload attempt was aborted and cannot be resumed. The reason was:"
```
Comment (mephenor) [inline ucs/domain/interfaces/inbound/upload_service.py]:
```suggestion
            + " The upload attempt was aborted and cannot be resumed. The reason was: "
```

File: ucs/domain/interfaces/inbound/upload_service.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        super().__init__(message)
+
+
+class UploadCancelError(RuntimeError):
+    """Thrown when the cancelling of an upload attempt failed."""
+
+    def __init__(self, *, upload_id: str):
+        self.upload_id = upload_id
+        self.possible_reason = (
+            "An ongoing part upload might be a reason. Please complete all part upload"
```
Comment (mephenor) [inline ucs/domain/interfaces/inbound/upload_service.py]:
```suggestion
            "An ongoing part upload might be a reason. Please complete all part uploads"
```

Comment (mephenor) [reply ucs/domain/interfaces/inbound/upload_service.py]: This seems weird... shouldn't we take care of cancelling the multipart uploads in this case?

**Review by mephenor — APPROVED** [review]
Some small suggestions (mostly typos) and questions here and there.
All in all, looks very nice, though I can't guarantee that I did not miss some minor issues, as this was quite a bit of content.

Just one bit: 
For ucs/domain/interfaces/inbound/upload_service.py wouldn't it be better to factor out the errors and just keep the interface there? Would be consistent with how we've done that in other places

File: .static_files [hunk 18]
```diff
@@ -36,6 +36,3 @@ example_data/README.md
 LICENSE
 requirements.txt
 setup.py
-
-.static_files
```
Comment (MoritzHahn1337) [inline .static_files:40]: `.static_files` and `mandatory_files` should at least be in the mandatory files list.

File: tests/fixtures/s3.py (hunk 1 above)
Comment (MoritzHahn1337) [reply tests/fixtures/s3.py:55]: I agree. We shouldn't hardcode it here - at least put it as a constant

**Review by MoritzHahn1337 — APPROVED** [review]
